{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "color-resource",
   "metadata": {},
   "source": [
    "# Check Goat coordinates\n",
    "## Check goat coordinates with data provided by Goat genome project\n",
    "Test coordinates received from goat genome consortium\n",
    "## Working on CHI1\n",
    "Setting up stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "announced-possibility",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import json\n",
    "import pprint\n",
    "import random\n",
    "import pandas\n",
    "import functools\n",
    "\n",
    "from collections import namedtuple, defaultdict\n",
    "from ensemblrest import EnsemblRest\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "from src.data.common import WORKING_ASSEMBLIES\n",
    "from src.features.utils import get_project_dir, text_or_gzip_open, sanitize\n",
    "from src.features.smarterdb import global_connection, VariantGoat\n",
    "\n",
    "# Always tell NCBI who you are\n",
    "Entrez.email = \"paolo.cozzi@ibba.cnr.it\"\n",
    "\n",
    "# define ensemblrest helper\n",
    "ensRest = EnsemblRest()\n",
    "\n",
    "# define the source data file\n",
    "data_dir = get_project_dir() / \"data/external/GOA/CONSORTIUM\"\n",
    "capri_all_data = data_dir / \"capri4dbsnp-base-CHI-ARS-OAR-UMD.csv.gz\"\n",
    "\n",
    "# make a connection to smarter database\n",
    "global_connection()\n",
    "\n",
    "# get the smarter assembly I want to test\n",
    "CHI1 = WORKING_ASSEMBLIES['CHI1']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sharing-glance",
   "metadata": {},
   "source": [
    "Defining some useful functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "demonstrated-ceramic",
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=1000)\n",
    "def get_chr_from_ncbi(chrom_name: str):\n",
    "    if chrom_name == 'null':\n",
    "        return None\n",
    "    \n",
    "    if chrom_name.isdigit():\n",
    "        return chrom_name\n",
    "    \n",
    "    # get data from NCBI\n",
    "    with Entrez.efetch(db=\"nucleotide\", id=chrom_name, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        \n",
    "    return record.features[0].qualifiers[\"chromosome\"][0]\n",
    "\n",
    "\n",
    "def check_coordinates_chi1(record, variant):\n",
    "    check = True\n",
    "    location = variant.locations[0]\n",
    "    \n",
    "    # ['chi_1_0_chr', 'chi_1_0_pos', 'chi_1_0_strand']\n",
    "    if record.chi_1_0_chr != location.chrom:\n",
    "        check = False\n",
    "        \n",
    "    if record.chi_1_0_pos != location.position:\n",
    "        check = False\n",
    "        \n",
    "    return check\n",
    "\n",
    "\n",
    "def search_mismatches(datafile, smarter_assembly, check_coordinates):    \n",
    "    # track mismatches in positions\n",
    "    mismatches = []\n",
    "\n",
    "    # track missing sps\n",
    "    missing = []\n",
    "\n",
    "    with text_or_gzip_open(datafile) as handle:\n",
    "        reader = csv.reader(handle, delimiter=\"\\t\")\n",
    "        header = next(reader)\n",
    "        header = [sanitize(col) for col in header]\n",
    "        Record = namedtuple(\"Record\", header)\n",
    "        for line in reader:\n",
    "            # fix position columns\n",
    "            for field in filter(lambda field: 'pos' in field, header):\n",
    "                try:\n",
    "                    idx = header.index(field)\n",
    "                    line[idx] = int(line[idx])\n",
    "\n",
    "                except ValueError as exc:\n",
    "                    continue\n",
    "\n",
    "            # fix chi1 chrom names\n",
    "            idx = header.index('chi_1_0_chr')\n",
    "            line[idx] = get_chr_from_ncbi(line[idx])\n",
    "\n",
    "            # fix ars1 chrom_name\n",
    "            idx = header.index('ars1_chr')\n",
    "            line[idx] = get_chr_from_ncbi(line[idx])\n",
    "            \n",
    "            # make a record from csv line\n",
    "            record = Record._make(line)\n",
    "\n",
    "            # get variant and chosen location from database\n",
    "            try:\n",
    "                variant = VariantGoat.objects(name=record.locus_name).fields(\n",
    "                    pk=0, rs_id=1, name=1, elemMatch__locations=smarter_assembly._asdict()).get()\n",
    "            except VariantGoat.DoesNotExist:\n",
    "                # print(f\"Couldn't find variant {record.locus_name}\")\n",
    "                missing.append(record)\n",
    "                continue\n",
    "\n",
    "            # check data\n",
    "            if not check_coordinates(record, variant):\n",
    "                # print(f\"Variant is different: {record} <> {variant}, {variant.locations[0]}\")\n",
    "                mismatches.append([record, variant])\n",
    "\n",
    "    # print(record)\n",
    "    # pprint.pprint(json.loads(variant.to_json()))\n",
    "    print(f\"There are {len(mismatches)} mismatches\")\n",
    "    print(f\"{len(missing)} SNPs are not in database\")\n",
    "    \n",
    "    return mismatches, missing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "medium-season",
   "metadata": {},
   "source": [
    "Now check *CHI1* coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "nominated-nickname",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1420 mismatches\n",
      "6653 SNPs are not in database\n"
     ]
    }
   ],
   "source": [
    "mismatches_chi1, missing_chi1 = search_mismatches(capri_all_data, CHI1, check_coordinates_chi1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eight-photographer",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(a=42, version=2)\n",
    "selected_chi1 = random.sample(mismatches_chi1, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "actual-taste",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_name</th>\n",
       "      <th>rs_</th>\n",
       "      <th>chi_1_0_chr</th>\n",
       "      <th>chi_1_0_pos</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>smarter_chrom</th>\n",
       "      <th>smarter_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>snp53417-scaffold807-74773</td>\n",
       "      <td>rs268285051</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>74773</td>\n",
       "      <td>rs268285051</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>snp10135-scaffold1361-44576</td>\n",
       "      <td>rs268243008</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>44576</td>\n",
       "      <td>rs268243008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>snp2042-scaffold1061-20259</td>\n",
       "      <td>rs268235116</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>20259</td>\n",
       "      <td>rs268235116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>snp21860-scaffold2144-53078</td>\n",
       "      <td>rs268254327</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>53078</td>\n",
       "      <td>rs268254327</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>snp19559-scaffold1971-21103</td>\n",
       "      <td>rs268252089</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>21103</td>\n",
       "      <td>rs268252089</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>snp18430-scaffold1867-78532</td>\n",
       "      <td>rs268251028</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>78532</td>\n",
       "      <td>rs268251028</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>snp13189-scaffold1505-57602</td>\n",
       "      <td>rs268245970</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>57602</td>\n",
       "      <td>rs268245970</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>snp6785-scaffold1249-70584</td>\n",
       "      <td>rs268239737</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>70584</td>\n",
       "      <td>rs268239737</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>snp59632-scaffold985-45095</td>\n",
       "      <td>rs268291047</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>45095</td>\n",
       "      <td>rs268291047</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>snp37483-scaffold4582-6164</td>\n",
       "      <td>rs268269564</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6164</td>\n",
       "      <td>rs268269564</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>snp6280-scaffold1221-917793</td>\n",
       "      <td>rs268239253</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>917793</td>\n",
       "      <td>rs268239253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>snp43513-scaffold58-61114</td>\n",
       "      <td>rs268275399</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>61114</td>\n",
       "      <td>rs268275399</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>snp27762-scaffold2961-47017</td>\n",
       "      <td>rs268260077</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>47017</td>\n",
       "      <td>rs268260077</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>snp3259-scaffold10997-4300</td>\n",
       "      <td>rs268236298</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>4300</td>\n",
       "      <td>rs268236298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>snp2897-scaffold1087-335720</td>\n",
       "      <td>rs268235947</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>335720</td>\n",
       "      <td>rs268235947</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snp6295-scaffold1221-1449031</td>\n",
       "      <td>rs268239266</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1449031</td>\n",
       "      <td>rs268239266</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>snp18103-scaffold185-12751132</td>\n",
       "      <td>rs268250717</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>5818</td>\n",
       "      <td>rs268250717</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>snp19056-scaffold1913-6125</td>\n",
       "      <td>rs268251607</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6125</td>\n",
       "      <td>rs268251607</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>snp33811-scaffold398-1222054</td>\n",
       "      <td>rs268265954</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1222054</td>\n",
       "      <td>rs268265954</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>snp47214-scaffold6615-3768</td>\n",
       "      <td>rs268279011</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3768</td>\n",
       "      <td>rs268279011</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       locus_name          rs_ chi_1_0_chr  chi_1_0_pos  \\\n",
       "0      snp53417-scaffold807-74773  rs268285051     Unknown        74773   \n",
       "1     snp10135-scaffold1361-44576  rs268243008     Unknown        44576   \n",
       "2      snp2042-scaffold1061-20259  rs268235116     Unknown        20259   \n",
       "3     snp21860-scaffold2144-53078  rs268254327     Unknown        53078   \n",
       "4     snp19559-scaffold1971-21103  rs268252089     Unknown        21103   \n",
       "5     snp18430-scaffold1867-78532  rs268251028     Unknown        78532   \n",
       "6     snp13189-scaffold1505-57602  rs268245970     Unknown        57602   \n",
       "7      snp6785-scaffold1249-70584  rs268239737     Unknown        70584   \n",
       "8      snp59632-scaffold985-45095  rs268291047     Unknown        45095   \n",
       "9      snp37483-scaffold4582-6164  rs268269564     Unknown         6164   \n",
       "10    snp6280-scaffold1221-917793  rs268239253     Unknown       917793   \n",
       "11      snp43513-scaffold58-61114  rs268275399     Unknown        61114   \n",
       "12    snp27762-scaffold2961-47017  rs268260077     Unknown        47017   \n",
       "13     snp3259-scaffold10997-4300  rs268236298     Unknown         4300   \n",
       "14    snp2897-scaffold1087-335720  rs268235947     Unknown       335720   \n",
       "15   snp6295-scaffold1221-1449031  rs268239266     Unknown      1449031   \n",
       "16  snp18103-scaffold185-12751132  rs268250717     Unknown         5818   \n",
       "17     snp19056-scaffold1913-6125  rs268251607     Unknown         6125   \n",
       "18   snp33811-scaffold398-1222054  rs268265954     Unknown      1222054   \n",
       "19     snp47214-scaffold6615-3768  rs268279011     Unknown         3768   \n",
       "\n",
       "          rs_id smarter_chrom  smarter_pos  \n",
       "0   rs268285051             0            0  \n",
       "1   rs268243008             0            0  \n",
       "2   rs268235116             0            0  \n",
       "3   rs268254327             0            0  \n",
       "4   rs268252089             0            0  \n",
       "5   rs268251028             0            0  \n",
       "6   rs268245970             0            0  \n",
       "7   rs268239737             0            0  \n",
       "8   rs268291047             0            0  \n",
       "9   rs268269564             0            0  \n",
       "10  rs268239253             0            0  \n",
       "11  rs268275399             0            0  \n",
       "12  rs268260077             0            0  \n",
       "13  rs268236298             0            0  \n",
       "14  rs268235947             0            0  \n",
       "15  rs268239266             0            0  \n",
       "16  rs268250717             0            0  \n",
       "17  rs268251607             0            0  \n",
       "18  rs268265954             0            0  \n",
       "19  rs268279011             0            0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = defaultdict(list)\n",
    "\n",
    "for record, variant in selected_chi1:\n",
    "    data[\"locus_name\"].append(record.locus_name)\n",
    "    data[\"rs_\"].append(record.rs_)\n",
    "    data[\"chi_1_0_chr\"].append(record.chi_1_0_chr)\n",
    "    data[\"chi_1_0_pos\"].append(record.chi_1_0_pos)\n",
    "    data[\"rs_id\"].append(variant.rs_id)\n",
    "    data[\"smarter_chrom\"].append(variant.locations[0].chrom)\n",
    "    data[\"smarter_pos\"].append(variant.locations[0].position)\n",
    "\n",
    "pandas.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "needed-nutrition",
   "metadata": {},
   "source": [
    "Coordinates seems to match with database for *CHI1* assembly. Differences are in contigs (unknown chromosomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "geographic-daisy",
   "metadata": {},
   "source": [
    "## Working with ARS1 coordinates\n",
    "Checking latest coordinates version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bound-holiday",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the smarter assembly I want to test\n",
    "ARS1 = WORKING_ASSEMBLIES['ARS1']\n",
    "\n",
    "# defining a new function to test coordinates\n",
    "def check_coordinates_ars1(record, variant):\n",
    "    check = True\n",
    "    location = variant.locations[0]\n",
    "    \n",
    "    # ['ars1_chr', 'ars1_pos', 'ars1_strand']\n",
    "    if record.ars1_chr != location.chrom:\n",
    "        check = False\n",
    "        \n",
    "    if record.ars1_pos != location.position:\n",
    "        check = False\n",
    "        \n",
    "    return check"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "optimum-haven",
   "metadata": {},
   "source": [
    "Search mismatches in *ARS1* assembly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "treated-resident",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1184 mismatches\n",
      "6653 SNPs are not in database\n"
     ]
    }
   ],
   "source": [
    "mismatches_ars1, missing_ars1 = search_mismatches(capri_all_data, ARS1, check_coordinates_ars1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "extended-metro",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(a=42, version=2)\n",
    "selected_ars1 = random.sample(mismatches_ars1, 20)\n",
    "\n",
    "# get rs_id from selected items\n",
    "selected_rsID = [variant.rs_id for (record, variant) in selected_ars1 if variant.rs_id]\n",
    "\n",
    "# get informations from ensembl\n",
    "result_ars1 = ensRest.getVariationByMultipleIds(ids=selected_rsID, species=\"capra_hircus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "southeast-nicholas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry</th>\n",
       "      <th>rs_</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>smarter_chrom</th>\n",
       "      <th>smarter_pos</th>\n",
       "      <th>ensembl_chrom</th>\n",
       "      <th>ensembl_pos</th>\n",
       "      <th>ensembl_failed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>snp10978-scaffold13932-19256</td>\n",
       "      <td>rs268243822</td>\n",
       "      <td>18</td>\n",
       "      <td>64350718</td>\n",
       "      <td>rs268243822</td>\n",
       "      <td>18</td>\n",
       "      <td>64495025</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>snp2720-scaffold1079-655080</td>\n",
       "      <td>rs268235776</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>20183</td>\n",
       "      <td>rs268235776</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>20183</td>\n",
       "      <td>7</td>\n",
       "      <td>92001864.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>snp24769-scaffold2530-14952</td>\n",
       "      <td>rs268257162</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>36932</td>\n",
       "      <td>rs268257162</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>9829</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>snp22886-scaffold228-4649518</td>\n",
       "      <td>rs268255332</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>372155</td>\n",
       "      <td>rs268255332</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>372155</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>snp21494-scaffold21-244274</td>\n",
       "      <td>rs268253976</td>\n",
       "      <td>14</td>\n",
       "      <td>-1</td>\n",
       "      <td>rs268253976</td>\n",
       "      <td>null</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>snp13706-scaffold153-960292</td>\n",
       "      <td>rs268246476</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>858874</td>\n",
       "      <td>rs268246476</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>858874</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>snp8812-scaffold1316-38259</td>\n",
       "      <td>rs268241723</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>160485</td>\n",
       "      <td>rs268241723</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>160485</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>snp58664-scaffold956-522857</td>\n",
       "      <td>rs268290160</td>\n",
       "      <td>13</td>\n",
       "      <td>71181569</td>\n",
       "      <td>rs268290160</td>\n",
       "      <td>13</td>\n",
       "      <td>71187835</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant maps to more than one genomic location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>snp7608-scaffold1271-347413</td>\n",
       "      <td>rs268240542</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>146208</td>\n",
       "      <td>rs268240542</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>146208</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>snp37951-scaffold47-1337018</td>\n",
       "      <td>rs268270029</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>57698</td>\n",
       "      <td>rs268270029</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>57698</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>snp2897-scaffold1087-335720</td>\n",
       "      <td>rs268235947</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>167713</td>\n",
       "      <td>rs268235947</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>167713</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>snp2893-scaffold1087-42518</td>\n",
       "      <td>rs268235943</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>36835</td>\n",
       "      <td>rs268235943</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>36835</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>snp8034-scaffold1291-127632</td>\n",
       "      <td>rs268240961</td>\n",
       "      <td>None</td>\n",
       "      <td>null</td>\n",
       "      <td>rs268240961</td>\n",
       "      <td>null</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>snp20479-scaffold202-4950867</td>\n",
       "      <td>rs268252989</td>\n",
       "      <td>2</td>\n",
       "      <td>100856547</td>\n",
       "      <td>rs268252989</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>8887</td>\n",
       "      <td>2</td>\n",
       "      <td>100856547.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>snp22223-scaffold2195-17488</td>\n",
       "      <td>rs268254680</td>\n",
       "      <td>3</td>\n",
       "      <td>111420771</td>\n",
       "      <td>rs268254680</td>\n",
       "      <td>3</td>\n",
       "      <td>111576031</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snp52657-scaffold789-1022</td>\n",
       "      <td>rs268284313</td>\n",
       "      <td>X</td>\n",
       "      <td>63743447</td>\n",
       "      <td>rs268284313</td>\n",
       "      <td>null</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>snp2772-scaffold108-589640</td>\n",
       "      <td>rs268235826</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7740</td>\n",
       "      <td>rs268235826</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>7740</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>snp931-scaffold1025-772629</td>\n",
       "      <td>rs268291431</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>12794</td>\n",
       "      <td>rs268291431</td>\n",
       "      <td>6</td>\n",
       "      <td>98497277</td>\n",
       "      <td>6</td>\n",
       "      <td>98497277.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>snp19041-scaffold1912-370148</td>\n",
       "      <td>rs268251592</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>13197</td>\n",
       "      <td>rs268251592</td>\n",
       "      <td>22</td>\n",
       "      <td>54336948</td>\n",
       "      <td>22</td>\n",
       "      <td>54336948.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>snp37946-scaffold47-1110486</td>\n",
       "      <td>rs268270024</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>289789</td>\n",
       "      <td>rs268270024</td>\n",
       "      <td>0.NW_scaffold</td>\n",
       "      <td>289789</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant does not map to the genome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           entry          rs_    chrom        pos  \\\n",
       "0   snp10978-scaffold13932-19256  rs268243822       18   64350718   \n",
       "1    snp2720-scaffold1079-655080  rs268235776  Unknown      20183   \n",
       "2    snp24769-scaffold2530-14952  rs268257162  Unknown      36932   \n",
       "3   snp22886-scaffold228-4649518  rs268255332  Unknown     372155   \n",
       "4     snp21494-scaffold21-244274  rs268253976       14         -1   \n",
       "5    snp13706-scaffold153-960292  rs268246476  Unknown     858874   \n",
       "6     snp8812-scaffold1316-38259  rs268241723  Unknown     160485   \n",
       "7    snp58664-scaffold956-522857  rs268290160       13   71181569   \n",
       "8    snp7608-scaffold1271-347413  rs268240542  Unknown     146208   \n",
       "9    snp37951-scaffold47-1337018  rs268270029  Unknown      57698   \n",
       "10   snp2897-scaffold1087-335720  rs268235947  Unknown     167713   \n",
       "11    snp2893-scaffold1087-42518  rs268235943  Unknown      36835   \n",
       "12   snp8034-scaffold1291-127632  rs268240961     None       null   \n",
       "13  snp20479-scaffold202-4950867  rs268252989        2  100856547   \n",
       "14   snp22223-scaffold2195-17488  rs268254680        3  111420771   \n",
       "15     snp52657-scaffold789-1022  rs268284313        X   63743447   \n",
       "16    snp2772-scaffold108-589640  rs268235826  Unknown       7740   \n",
       "17    snp931-scaffold1025-772629  rs268291431  Unknown      12794   \n",
       "18  snp19041-scaffold1912-370148  rs268251592  Unknown      13197   \n",
       "19   snp37946-scaffold47-1110486  rs268270024  Unknown     289789   \n",
       "\n",
       "          rs_id  smarter_chrom  smarter_pos ensembl_chrom  ensembl_pos  \\\n",
       "0   rs268243822             18     64495025          None          NaN   \n",
       "1   rs268235776  0.NW_scaffold        20183             7   92001864.0   \n",
       "2   rs268257162  0.NW_scaffold         9829          None          NaN   \n",
       "3   rs268255332  0.NW_scaffold       372155          None          NaN   \n",
       "4   rs268253976           null            0          None          NaN   \n",
       "5   rs268246476  0.NW_scaffold       858874          None          NaN   \n",
       "6   rs268241723  0.NW_scaffold       160485          None          NaN   \n",
       "7   rs268290160             13     71187835          None          NaN   \n",
       "8   rs268240542  0.NW_scaffold       146208          None          NaN   \n",
       "9   rs268270029  0.NW_scaffold        57698          None          NaN   \n",
       "10  rs268235947  0.NW_scaffold       167713          None          NaN   \n",
       "11  rs268235943  0.NW_scaffold        36835          None          NaN   \n",
       "12  rs268240961           null            0          None          NaN   \n",
       "13  rs268252989  0.NW_scaffold         8887             2  100856547.0   \n",
       "14  rs268254680              3    111576031          None          NaN   \n",
       "15  rs268284313           null            0          None          NaN   \n",
       "16  rs268235826  0.NW_scaffold         7740          None          NaN   \n",
       "17  rs268291431              6     98497277             6   98497277.0   \n",
       "18  rs268251592             22     54336948            22   54336948.0   \n",
       "19  rs268270024  0.NW_scaffold       289789          None          NaN   \n",
       "\n",
       "                                    ensembl_failed  \n",
       "0               Variant does not map to the genome  \n",
       "1                                             None  \n",
       "2               Variant does not map to the genome  \n",
       "3               Variant does not map to the genome  \n",
       "4               Variant does not map to the genome  \n",
       "5               Variant does not map to the genome  \n",
       "6               Variant does not map to the genome  \n",
       "7   Variant maps to more than one genomic location  \n",
       "8               Variant does not map to the genome  \n",
       "9               Variant does not map to the genome  \n",
       "10              Variant does not map to the genome  \n",
       "11              Variant does not map to the genome  \n",
       "12              Variant does not map to the genome  \n",
       "13                                            None  \n",
       "14              Variant does not map to the genome  \n",
       "15              Variant does not map to the genome  \n",
       "16              Variant does not map to the genome  \n",
       "17                                            None  \n",
       "18                                            None  \n",
       "19              Variant does not map to the genome  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def to_pd_table(selected, result):\n",
    "    data = defaultdict(list)\n",
    "\n",
    "    for record, variant in selected:\n",
    "        data[\"entry\"].append(record.locus_name)\n",
    "        data[\"rs_\"].append(record.rs_)\n",
    "        data[\"chrom\"].append(record.ars1_chr)\n",
    "        data[\"pos\"].append(record.ars1_pos)\n",
    "        data[\"rs_id\"].append(variant.rs_id)\n",
    "        data[\"smarter_chrom\"].append(variant.locations[0].chrom)\n",
    "        data[\"smarter_pos\"].append(variant.locations[0].position)\n",
    "\n",
    "        if variant.rs_id and not 'failed' in result[variant.rs_id]:\n",
    "            data[\"ensembl_chrom\"].append(result[variant.rs_id]['mappings'][0]['seq_region_name'])\n",
    "            data[\"ensembl_pos\"].append(result[variant.rs_id]['mappings'][0]['start'])\n",
    "        \n",
    "        else:\n",
    "            data[\"ensembl_chrom\"].append(None)\n",
    "            data[\"ensembl_pos\"].append(None)\n",
    "        \n",
    "        if 'failed' in result[variant.rs_id]:\n",
    "            data[\"ensembl_failed\"].append(result[variant.rs_id]['failed'])\n",
    "        \n",
    "        else:\n",
    "            data[\"ensembl_failed\"].append(None)\n",
    "\n",
    "    return pandas.DataFrame.from_dict(data)\n",
    "\n",
    "to_pd_table(selected_ars1, result_ars1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collected-monaco",
   "metadata": {},
   "source": [
    "There are few mismatches from received coordinates and the data I have in the database. However, the smarter data comes from the latest chip so I belive I can trust it. There are some considerations to made about SNP placement: I thick I need to cross-check with ensembl and try to align the probe myself to understand if there are issues on chip placement"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
