{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "17d31879-3132-4e3d-8ec0-025f3254540c",
   "metadata": {},
   "source": [
    "# ISHEEP dataset\n",
    "Attempt to explore the ISHEEP dataset to understand if it can be imported into *SMARTER-database*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "82b25d5e-c783-460e-b85e-e86f18e06141",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import functools\n",
    "import requests\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "from plinkio import plinkfile\n",
    "from tqdm.notebook import tqdm\n",
    "from geopy.geocoders import Nominatim\n",
    "\n",
    "from src.features.utils import get_project_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87799699-7251-436e-ba88-231b75cb2428",
   "metadata": {},
   "source": [
    "Define some useful functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1d9668e-199e-4060-846f-9c540156afb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "geolocator = Nominatim(user_agent=\"SMARTER-database\")\n",
    "\n",
    "\n",
    "@functools.cache\n",
    "def get_location(sample_location):\n",
    "    location = geolocator.geocode(sample_location)\n",
    "    \n",
    "    if not location:\n",
    "        return None, None, None\n",
    "        \n",
    "    return location.address, location.latitude, location.longitude\n",
    "\n",
    "\n",
    "# https://stackoverflow.com/a/20558779\n",
    "def stripNone(data):\n",
    "    if isinstance(data, dict):\n",
    "        return {k:stripNone(v) for k, v in data.items() if k is not None and v is not None}\n",
    "    elif isinstance(data, list):\n",
    "        return [stripNone(item) for item in data if item is not None]\n",
    "    elif isinstance(data, tuple):\n",
    "        return tuple(stripNone(item) for item in data if item is not None)\n",
    "    elif isinstance(data, set):\n",
    "        return {stripNone(item) for item in data if item is not None}\n",
    "    else:\n",
    "        return data\n",
    "\n",
    "    \n",
    "@functools.cache\n",
    "def fetch_cncb_biosample(biosample_id):\n",
    "    response = requests.get(f\"https://ngdc.cncb.ac.cn/gwh/api/public/bioSample/{biosample_id}\")\n",
    "    \n",
    "    if response.status_code != 200:\n",
    "        return None\n",
    "    \n",
    "    data = response.json()\n",
    "    data = stripNone(data)\n",
    "    \n",
    "    if \"submitter\" in data:\n",
    "        _ = data.pop(\"submitter\")\n",
    "    \n",
    "    if \"taxon\" in data:\n",
    "        _ = data.pop(\"taxon\")\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a599bb14-d5ae-45e5-a928-6ea14d697237",
   "metadata": {},
   "source": [
    "Read all data from *ISHEEP dataset*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e4e09ec3-7ece-4709-b3a4-5e585feddc78",
   "metadata": {},
   "outputs": [],
   "source": [
    "tablefile = str(get_project_dir() / \"data/external/SHE/ISHEEP/Table_1_fixed.xlsx\")\n",
    "metadata = pd.read_excel(tablefile, sheet_name=\"Table S1\", skiprows=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b33d2f09-e178-4fd4-baf7-4f135541dc4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AUMEHM01</td>\n",
       "      <td>ERS1460923</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.95</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUMEHM02</td>\n",
       "      <td>ERS1460904</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.87</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AUMEHM03</td>\n",
       "      <td>ERS1460978</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.42</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AUMEHM04</td>\n",
       "      <td>ERS1461020</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.45</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AUMEPM05</td>\n",
       "      <td>ERS1461055</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Polled</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.46</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample ID Biosample ID Bioproject ID     Species          Breed   Sex  \\\n",
       "0  AUMEHM01   ERS1460923     SRP078481  Ovis aries  Merino Horned  male   \n",
       "1  AUMEHM02   ERS1460904     SRP078481  Ovis aries  Merino Horned  male   \n",
       "2  AUMEHM03   ERS1460978     SRP078481  Ovis aries  Merino Horned  male   \n",
       "3  AUMEHM04   ERS1461020     SRP078481  Ovis aries  Merino Horned  male   \n",
       "4  AUMEPM05   ERS1461055     SRP078481  Ovis aries  Merino Polled  male   \n",
       "\n",
       "     Sample location Material           Technology Assay type Coverage  \\\n",
       "0  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.95   \n",
       "1  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.87   \n",
       "2  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.42   \n",
       "3  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.45   \n",
       "4  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.46   \n",
       "\n",
       "  data resource  \n",
       "0           GVM  \n",
       "1           GVM  \n",
       "2           GVM  \n",
       "3           GVM  \n",
       "4           GVM  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a2c58b-843b-4b5b-85b8-70a4335f4012",
   "metadata": {},
   "source": [
    "Those are the different data types present in the isheep dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7e791446-e544-4d15-a679-be157c26bfda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['WGS', '50K chip', '600k chip'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"Assay type\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb4534e-2520-40a4-abd3-e98627f16890",
   "metadata": {},
   "source": [
    "# 50K data\n",
    "ok, let's subset the 50k dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa8387a6-793b-4c8e-a838-d953bda0cdff",
   "metadata": {},
   "outputs": [],
   "source": [
    "isheep_50K = metadata[metadata[\"Assay type\"] == \"50K chip\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c06cbdd5-fd64-43a9-8d15-f5f046d2368a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 1512 entries, 355 to 1866\n",
      "Data columns (total 12 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   Sample ID        1512 non-null   object\n",
      " 1   Biosample ID     1512 non-null   object\n",
      " 2   Bioproject ID    1512 non-null   object\n",
      " 3   Species          1512 non-null   object\n",
      " 4   Breed            1512 non-null   object\n",
      " 5   Sex              1512 non-null   object\n",
      " 6   Sample location  1512 non-null   object\n",
      " 7   Material         1512 non-null   object\n",
      " 8   Technology       1512 non-null   object\n",
      " 9   Assay type       1512 non-null   object\n",
      " 10  Coverage         0 non-null      object\n",
      " 11  data resource    1512 non-null   object\n",
      "dtypes: object(12)\n",
      "memory usage: 153.6+ KB\n"
     ]
    }
   ],
   "source": [
    "isheep_50K.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d929ef7-a44c-4db2-a928-16ada809c911",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>ARG10</td>\n",
       "      <td>SAMC060792</td>\n",
       "      <td>PRJCA001252</td>\n",
       "      <td>Ovis ammon</td>\n",
       "      <td>Argali</td>\n",
       "      <td>-</td>\n",
       "      <td>Xinjiang, China</td>\n",
       "      <td>Ear</td>\n",
       "      <td>Illumina Ovine SNP50K (54,241 SNPs) BeadChip</td>\n",
       "      <td>50K chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zhao et al., 2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>ARG11</td>\n",
       "      <td>SAMC060793</td>\n",
       "      <td>PRJCA001252</td>\n",
       "      <td>Ovis ammon</td>\n",
       "      <td>Argali</td>\n",
       "      <td>-</td>\n",
       "      <td>Xinjiang, China</td>\n",
       "      <td>Ear</td>\n",
       "      <td>Illumina Ovine SNP50K (54,241 SNPs) BeadChip</td>\n",
       "      <td>50K chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zhao et al., 2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>ARG9</td>\n",
       "      <td>SAMC060794</td>\n",
       "      <td>PRJCA001252</td>\n",
       "      <td>Ovis ammon</td>\n",
       "      <td>Argali</td>\n",
       "      <td>-</td>\n",
       "      <td>Xinjiang, China</td>\n",
       "      <td>Ear</td>\n",
       "      <td>Illumina Ovine SNP50K (54,241 SNPs) BeadChip</td>\n",
       "      <td>50K chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zhao et al., 2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>ARG6</td>\n",
       "      <td>SAMC060795</td>\n",
       "      <td>PRJCA001252</td>\n",
       "      <td>Ovis ammon</td>\n",
       "      <td>Argali</td>\n",
       "      <td>-</td>\n",
       "      <td>Xinjiang, China</td>\n",
       "      <td>Ear</td>\n",
       "      <td>Illumina Ovine SNP50K (54,241 SNPs) BeadChip</td>\n",
       "      <td>50K chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zhao et al., 2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>ARG8</td>\n",
       "      <td>SAMC060796</td>\n",
       "      <td>PRJCA001252</td>\n",
       "      <td>Ovis ammon</td>\n",
       "      <td>Argali</td>\n",
       "      <td>-</td>\n",
       "      <td>Xinjiang, China</td>\n",
       "      <td>Ear</td>\n",
       "      <td>Illumina Ovine SNP50K (54,241 SNPs) BeadChip</td>\n",
       "      <td>50K chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zhao et al., 2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample ID Biosample ID Bioproject ID     Species    Breed Sex  \\\n",
       "355     ARG10   SAMC060792   PRJCA001252  Ovis ammon  Argali    -   \n",
       "356     ARG11   SAMC060793   PRJCA001252  Ovis ammon  Argali    -   \n",
       "357      ARG9   SAMC060794   PRJCA001252  Ovis ammon  Argali    -   \n",
       "358      ARG6   SAMC060795   PRJCA001252  Ovis ammon  Argali    -   \n",
       "359      ARG8   SAMC060796   PRJCA001252  Ovis ammon  Argali    -   \n",
       "\n",
       "     Sample location Material                                    Technology  \\\n",
       "355  Xinjiang, China      Ear  Illumina Ovine SNP50K (54,241 SNPs) BeadChip   \n",
       "356  Xinjiang, China      Ear  Illumina Ovine SNP50K (54,241 SNPs) BeadChip   \n",
       "357  Xinjiang, China      Ear  Illumina Ovine SNP50K (54,241 SNPs) BeadChip   \n",
       "358  Xinjiang, China      Ear  Illumina Ovine SNP50K (54,241 SNPs) BeadChip   \n",
       "359  Xinjiang, China      Ear  Illumina Ovine SNP50K (54,241 SNPs) BeadChip   \n",
       "\n",
       "    Assay type Coverage      data resource  \n",
       "355   50K chip      NaN  Zhao et al., 2017  \n",
       "356   50K chip      NaN  Zhao et al., 2017  \n",
       "357   50K chip      NaN  Zhao et al., 2017  \n",
       "358   50K chip      NaN  Zhao et al., 2017  \n",
       "359   50K chip      NaN  Zhao et al., 2017  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_50K.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dbafe2d-5599-4343-8fba-00550fac6331",
   "metadata": {},
   "source": [
    "There are species different from *Ovis Aries* in dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "09a7852f-8212-41d9-b604-65047e3250e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Ovis ammon', 'Ovis orientalis', 'Ovis aries'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_50K[\"Species\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62213ef5-385a-4b67-82f1-5e55618af06b",
   "metadata": {},
   "source": [
    "Get the different *breeds* from dataset. Need to remove the *sheep* from name, test for my SMARTER breeds, and deal with spaces and odd values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed3db845-5ad1-4ff7-ba5e-299b16f6678d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Argali ', 'mouflon ', 'Baerchuke sheep', 'Bashbay sheep',\n",
       "       'Celei black sheep', 'Diqing sheep', 'Guide Black Fur sheep',\n",
       "       'Guangling fat-tail sheep', 'Hulun Buir sheep', 'Hetian sheep',\n",
       "       'Hanzhong sheep', 'Jingzhong sheep', 'Kirghiz sheep',\n",
       "       'Lanzhou Large-tailed sheep', 'Lop sheep',\n",
       "       'Luzhong Mountain sheep', 'Minxian Black Fur sheep',\n",
       "       'Ninglang Black sheep', 'Sunite sheep', 'Shiping Gray sheep',\n",
       "       'Tan sheep', 'Tengchong sheep', 'Taihang Fur sheep',\n",
       "       'Turfan Black sheep', 'Tong sheep', 'Tashkurgan sheep',\n",
       "       'Lanping Black-bone sheep', 'Weining sheep', 'Wuranke sheep',\n",
       "       'Ujimqin sheep', 'Yecheng sheep', 'Yuxi Fat-tailed sheep',\n",
       "       'Zhaotong sheep', 'Altay sheep', 'Bayinbuluke sheep',\n",
       "       'Tibetan sheep', 'Duolang sheep', 'Kazakh sheep'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_50K[\"Breed\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6623901-7040-4705-a0c7-8159181ccbb8",
   "metadata": {},
   "source": [
    "Try to collect data from CNCB biosample or load data from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6317890-ac29-4029-8237-459f5ab23d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if Path(\"isheep_50K_biosample.pkl\").exists():\n",
    "    isheep_50K_biosample = pd.read_pickle(\"isheep_50K_biosample.pkl\")\n",
    "else:\n",
    "    biosample_data = [fetch_cncb_biosample(biosample_id) for biosample_id in tqdm(isheep_50K[\"Biosample ID\"])]\n",
    "    isheep_50K_biosample = pd.json_normalize(biosample_data)\n",
    "    isheep_50K_biosample.to_pickle(\"isheep_50K_biosample.pkl\")\n",
    "    isheep_50K_biosample.to_excel(\"isheep_50K_biosample.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d7f4bf-388f-4f0c-91d9-a36d42d752d5",
   "metadata": {},
   "source": [
    "Data collected from biosamples are not different from data collected from metadata. Try to open the processed plink file and check that all sample ids are represented"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d63f1018-7edf-4d2e-a586-4fe531e83274",
   "metadata": {},
   "outputs": [],
   "source": [
    "plinkio = plinkfile.open(str(get_project_dir() / \"data/external/SHE/ISHEEP/50K-all/50K-all\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8504cd68-270b-468a-8e71-d5efa2f32b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [sample.iid for sample in plinkio.get_samples()]\n",
    "for index, row in isheep_50K.iterrows():\n",
    "    if not row[\"Sample ID\"] in samples:\n",
    "        print(f\"Missing {row['Sample ID']}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1641978-2a2f-41f0-8784-982a6cd0b339",
   "metadata": {},
   "source": [
    "Sample names between metadata and plink file are identical. Try to derive sample location using geocoding:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "af4d33f6-cf33-42af-a10a-504328b791ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = isheep_50K[\"Sample location\"].apply(get_location)\n",
    "# isheep_50K[\"address\"] = test.apply(lambda x: x[0])\n",
    "# isheep_50K[\"latitude\"] = test.apply(lambda x: x[1])\n",
    "# isheep_50K[\"longitude\"] = test.apply(lambda x: x[2])\n",
    "# isheep_50K.head()\n",
    "isheep_50K.to_excel(\"isheep_50K.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77e60c2a-68f9-4d9a-aab1-392125e063ce",
   "metadata": {},
   "source": [
    "Despite is possible to derive coordinates from the location I have, I don't want to place a latitude and longitude in my database: if anyone want to consider coordinates from location, need to understand the coordinate precision level (I can't consider russia coordinate at the same level of a cina county)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a281e6f-302d-43a6-9caf-0de7a92ea998",
   "metadata": {},
   "source": [
    "## 600K data\n",
    "next, explore the 600k dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bda62b1b-7121-4d08-b0bf-3036e822991e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 911 entries, 1867 to 2777\n",
      "Data columns (total 12 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   Sample ID        911 non-null    object\n",
      " 1   Biosample ID     911 non-null    object\n",
      " 2   Bioproject ID    911 non-null    object\n",
      " 3   Species          911 non-null    object\n",
      " 4   Breed            911 non-null    object\n",
      " 5   Sex              911 non-null    object\n",
      " 6   Sample location  911 non-null    object\n",
      " 7   Material         911 non-null    object\n",
      " 8   Technology       911 non-null    object\n",
      " 9   Assay type       911 non-null    object\n",
      " 10  Coverage         0 non-null      object\n",
      " 11  data resource    911 non-null    object\n",
      "dtypes: object(12)\n",
      "memory usage: 92.5+ KB\n"
     ]
    }
   ],
   "source": [
    "isheep_600K = metadata[metadata[\"Assay type\"] == \"600k chip\"]\n",
    "isheep_600K.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "06f7824d-ecfe-466c-bd78-dce364b6354e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>DLS DLS355</td>\n",
       "      <td>SAMC062304</td>\n",
       "      <td>PRJCA001253</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Duolang sheep</td>\n",
       "      <td>female</td>\n",
       "      <td>Aksu region of Xinjiang province, China</td>\n",
       "      <td>ear</td>\n",
       "      <td>Illumina Ovine Infinium HD SNP BeadChip</td>\n",
       "      <td>600k chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gao et al., 2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868</th>\n",
       "      <td>DLS DLS.DLS249</td>\n",
       "      <td>SAMC062305</td>\n",
       "      <td>PRJCA001253</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Duolang sheep</td>\n",
       "      <td>female</td>\n",
       "      <td>Aksu region of Xinjiang province, China</td>\n",
       "      <td>ear</td>\n",
       "      <td>Illumina Ovine Infinium HD SNP BeadChip</td>\n",
       "      <td>600k chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gao et al., 2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869</th>\n",
       "      <td>DLS DLS.DLS302</td>\n",
       "      <td>SAMC062306</td>\n",
       "      <td>PRJCA001253</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Duolang sheep</td>\n",
       "      <td>male</td>\n",
       "      <td>Aksu region of Xinjiang province, China</td>\n",
       "      <td>ear</td>\n",
       "      <td>Illumina Ovine Infinium HD SNP BeadChip</td>\n",
       "      <td>600k chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gao et al., 2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1870</th>\n",
       "      <td>DLS DLS.DLS305</td>\n",
       "      <td>SAMC062307</td>\n",
       "      <td>PRJCA001253</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Duolang sheep</td>\n",
       "      <td>male</td>\n",
       "      <td>Aksu region of Xinjiang province, China</td>\n",
       "      <td>ear</td>\n",
       "      <td>Illumina Ovine Infinium HD SNP BeadChip</td>\n",
       "      <td>600k chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gao et al., 2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1871</th>\n",
       "      <td>DLS DLS.DLS309</td>\n",
       "      <td>SAMC062308</td>\n",
       "      <td>PRJCA001253</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Duolang sheep</td>\n",
       "      <td>male</td>\n",
       "      <td>Aksu region of Xinjiang province, China</td>\n",
       "      <td>ear</td>\n",
       "      <td>Illumina Ovine Infinium HD SNP BeadChip</td>\n",
       "      <td>600k chip</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gao et al., 2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample ID Biosample ID Bioproject ID     Species          Breed  \\\n",
       "1867      DLS DLS355   SAMC062304   PRJCA001253  Ovis aries  Duolang sheep   \n",
       "1868  DLS DLS.DLS249   SAMC062305   PRJCA001253  Ovis aries  Duolang sheep   \n",
       "1869  DLS DLS.DLS302   SAMC062306   PRJCA001253  Ovis aries  Duolang sheep   \n",
       "1870  DLS DLS.DLS305   SAMC062307   PRJCA001253  Ovis aries  Duolang sheep   \n",
       "1871  DLS DLS.DLS309   SAMC062308   PRJCA001253  Ovis aries  Duolang sheep   \n",
       "\n",
       "         Sex                          Sample location Material  \\\n",
       "1867  female  Aksu region of Xinjiang province, China      ear   \n",
       "1868  female  Aksu region of Xinjiang province, China      ear   \n",
       "1869    male  Aksu region of Xinjiang province, China      ear   \n",
       "1870    male  Aksu region of Xinjiang province, China      ear   \n",
       "1871    male  Aksu region of Xinjiang province, China      ear   \n",
       "\n",
       "                                   Technology Assay type Coverage  \\\n",
       "1867  Illumina Ovine Infinium HD SNP BeadChip  600k chip      NaN   \n",
       "1868  Illumina Ovine Infinium HD SNP BeadChip  600k chip      NaN   \n",
       "1869  Illumina Ovine Infinium HD SNP BeadChip  600k chip      NaN   \n",
       "1870  Illumina Ovine Infinium HD SNP BeadChip  600k chip      NaN   \n",
       "1871  Illumina Ovine Infinium HD SNP BeadChip  600k chip      NaN   \n",
       "\n",
       "         data resource  \n",
       "1867  Gao et al., 2018  \n",
       "1868  Gao et al., 2018  \n",
       "1869  Gao et al., 2018  \n",
       "1870  Gao et al., 2018  \n",
       "1871  Gao et al., 2018  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_600K.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7452f30c-692d-4a20-a9cf-72fd86672d7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Ovis aries'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_600K[\"Species\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6e1e403a-c116-41d1-a459-f7a7f3c2876a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Duolang sheep', 'Finnsheep', 'Texel sheep',\n",
       "       'Large Tailed Han sheep', 'Hu sheep', 'Wadi sheep',\n",
       "       'Icelandic sheep', 'Romanov sheep', 'Sishui Fur sheep',\n",
       "       'Small Tailed Han sheep'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_600K[\"Breed\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aadb2ba3-0b57-4492-811f-876989386c79",
   "metadata": {},
   "source": [
    "Try to collect data from CNCB biosample or load data from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d6234ea7-1a5f-445e-b5bc-337ebcd9d610",
   "metadata": {},
   "outputs": [],
   "source": [
    "if Path(\"isheep_600K_biosample.pkl\").exists():\n",
    "    isheep_600K_biosample = pd.read_pickle(\"isheep_600K_biosample.pkl\")\n",
    "else:\n",
    "    biosample_data = [fetch_cncb_biosample(biosample_id) for biosample_id in tqdm(isheep_600K[\"Biosample ID\"])]\n",
    "    isheep_600K_biosample = pd.json_normalize(biosample_data)\n",
    "    isheep_600K_biosample.to_pickle(\"isheep_600K_biosample.pkl\")\n",
    "    isheep_600K_biosample.to_excel(\"isheep_600K_biosample.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51a9fe4a-d941-46cf-bda4-372e34fbbc13",
   "metadata": {},
   "source": [
    "This time I can find coordinates in Biosamples in NE format (ie *38.37 N 77.23 E*). The sample name is the same I found in datafile. Need to define a sample alias equal to the sample name I find in plinkfile. Try to open the processed plink file and check that all sample ids are represented"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1c81a6e7-f8b8-481f-9daf-96a3db3b9b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "plinkio = plinkfile.open(str(get_project_dir() / \"data/external/SHE/ISHEEP/600K-all/600K-all\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56862b71-ea74-4566-becb-a364f429f29c",
   "metadata": {},
   "source": [
    "Check that sample names in plink file and metadata are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d1885f8b-f907-4c08-b1e6-69077f8c4b1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "121 samples name don't match between metadata and plink file\n"
     ]
    }
   ],
   "source": [
    "samples = [sample.iid for sample in plinkio.get_samples()]\n",
    "count = 0\n",
    "\n",
    "for index, row in isheep_600K.iterrows():\n",
    "    if not row[\"Sample ID\"] in samples:\n",
    "        # print(f\"Missing {row['Sample ID']}\")\n",
    "        count += 1\n",
    "        \n",
    "if count > 0:\n",
    "    print(f\"{count} samples name don't match between metadata and plink file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "53af9636-6fb8-496d-a08b-6b749b8ae69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.DataFrame({\"samples_id\": isheep_600K[\"Sample ID\"], \"samples_vcf\": samples})\n",
    "if not Path(\"samples_id-vcf-600K.xlsx\").exists():\n",
    "    test.to_excel(\"samples_id-vcf-600K.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8980bcd-25a3-49e4-9f7f-06ba9acac951",
   "metadata": {},
   "source": [
    "Save only 600K sample in a different file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e6dc7ae2-8cce-4bc9-a44d-8cf1835daebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "isheep_600K.to_excel(\"isheep_600K.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a44e3476-400c-47b2-a7c0-2e6b0ec7043b",
   "metadata": {},
   "source": [
    "## WGS dataset\n",
    "finally, get WGS metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "27455264-215d-4cb2-a2a8-fea81c8f6703",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 355 entries, 0 to 354\n",
      "Data columns (total 12 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   Sample ID        355 non-null    object\n",
      " 1   Biosample ID     355 non-null    object\n",
      " 2   Bioproject ID    355 non-null    object\n",
      " 3   Species          355 non-null    object\n",
      " 4   Breed            355 non-null    object\n",
      " 5   Sex              355 non-null    object\n",
      " 6   Sample location  355 non-null    object\n",
      " 7   Material         355 non-null    object\n",
      " 8   Technology       355 non-null    object\n",
      " 9   Assay type       355 non-null    object\n",
      " 10  Coverage         355 non-null    object\n",
      " 11  data resource    355 non-null    object\n",
      "dtypes: object(12)\n",
      "memory usage: 36.1+ KB\n"
     ]
    }
   ],
   "source": [
    "isheep_WGS = metadata[metadata[\"Assay type\"] == \"WGS\"].copy()\n",
    "isheep_WGS.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5ffef0b7-7a7a-4166-b437-6c2e1e584d51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AUMEHM01</td>\n",
       "      <td>ERS1460923</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.95</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUMEHM02</td>\n",
       "      <td>ERS1460904</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.87</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AUMEHM03</td>\n",
       "      <td>ERS1460978</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.42</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AUMEHM04</td>\n",
       "      <td>ERS1461020</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.45</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AUMEPM05</td>\n",
       "      <td>ERS1461055</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Polled</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.46</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample ID Biosample ID Bioproject ID     Species          Breed   Sex  \\\n",
       "0  AUMEHM01   ERS1460923     SRP078481  Ovis aries  Merino Horned  male   \n",
       "1  AUMEHM02   ERS1460904     SRP078481  Ovis aries  Merino Horned  male   \n",
       "2  AUMEHM03   ERS1460978     SRP078481  Ovis aries  Merino Horned  male   \n",
       "3  AUMEHM04   ERS1461020     SRP078481  Ovis aries  Merino Horned  male   \n",
       "4  AUMEPM05   ERS1461055     SRP078481  Ovis aries  Merino Polled  male   \n",
       "\n",
       "     Sample location Material           Technology Assay type Coverage  \\\n",
       "0  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.95   \n",
       "1  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.87   \n",
       "2  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.42   \n",
       "3  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.45   \n",
       "4  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.46   \n",
       "\n",
       "  data resource  \n",
       "0           GVM  \n",
       "1           GVM  \n",
       "2           GVM  \n",
       "3           GVM  \n",
       "4           GVM  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_WGS.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8744e6d5-f5c8-43bb-8959-a1f9cf553101",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 355 entries, 0 to 354\n",
      "Data columns (total 12 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   Sample ID        355 non-null    object\n",
      " 1   Biosample ID     355 non-null    object\n",
      " 2   Bioproject ID    355 non-null    object\n",
      " 3   Species          355 non-null    object\n",
      " 4   Breed            355 non-null    object\n",
      " 5   Sex              355 non-null    object\n",
      " 6   Sample location  355 non-null    object\n",
      " 7   Material         355 non-null    object\n",
      " 8   Technology       355 non-null    object\n",
      " 9   Assay type       355 non-null    object\n",
      " 10  Coverage         355 non-null    object\n",
      " 11  data resource    355 non-null    object\n",
      "dtypes: object(12)\n",
      "memory usage: 36.1+ KB\n"
     ]
    }
   ],
   "source": [
    "isheep_WGS = metadata[metadata[\"Assay type\"] == \"WGS\"]\n",
    "isheep_WGS.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5c20ba81-e882-431a-b0aa-a64188dae550",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AUMEHM01</td>\n",
       "      <td>ERS1460923</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.95</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUMEHM02</td>\n",
       "      <td>ERS1460904</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.87</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AUMEHM03</td>\n",
       "      <td>ERS1460978</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.42</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AUMEHM04</td>\n",
       "      <td>ERS1461020</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Horned</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.45</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AUMEPM05</td>\n",
       "      <td>ERS1461055</td>\n",
       "      <td>SRP078481</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Merino Polled</td>\n",
       "      <td>male</td>\n",
       "      <td>Oceania,Australia</td>\n",
       "      <td>Semen</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.46</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample ID Biosample ID Bioproject ID     Species          Breed   Sex  \\\n",
       "0  AUMEHM01   ERS1460923     SRP078481  Ovis aries  Merino Horned  male   \n",
       "1  AUMEHM02   ERS1460904     SRP078481  Ovis aries  Merino Horned  male   \n",
       "2  AUMEHM03   ERS1460978     SRP078481  Ovis aries  Merino Horned  male   \n",
       "3  AUMEHM04   ERS1461020     SRP078481  Ovis aries  Merino Horned  male   \n",
       "4  AUMEPM05   ERS1461055     SRP078481  Ovis aries  Merino Polled  male   \n",
       "\n",
       "     Sample location Material           Technology Assay type Coverage  \\\n",
       "0  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.95   \n",
       "1  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.87   \n",
       "2  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.42   \n",
       "3  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    16.45   \n",
       "4  Oceania,Australia    Semen  Illumina HiSeq 2500        WGS    14.46   \n",
       "\n",
       "  data resource  \n",
       "0           GVM  \n",
       "1           GVM  \n",
       "2           GVM  \n",
       "3           GVM  \n",
       "4           GVM  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_WGS.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9b974af3-1c0a-4a58-98fc-50f872a19092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Ovis aries', 'Ovis canadensis', 'Ovis dalli', 'Ovis ammon',\n",
       "       'Ovis orientalis'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_WGS[\"Species\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5c169996-ac98-43f1-a4c1-1bd6c54c6794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Merino Horned', 'Merino Polled', ' -', 'Katahdin',\n",
       "       'Tibetan sheep', 'Small tailed han sheep', 'Bighorn sheep',\n",
       "       'Afshari', 'Awassi', 'African White Dorper', 'Turkish Awassi',\n",
       "       'Brazilian Creole', 'Bangladeshi', 'Morada Nova', 'Santa Inês',\n",
       "       'Castellana', 'Cine Capari', 'Changthangi', 'Churra', 'Cheviot',\n",
       "       'Dollgellau Welsh Mountain', 'Ethiopian Menz', 'Finnsheep',\n",
       "       'Indian Garole', 'Gulf Coast native', 'Garut', 'Karya', 'Karakas',\n",
       "       'Meat Lacaune', 'Mmilk Lacaune', 'Merino', 'Norduz',\n",
       "       'Namaqua Afrikaner', 'Ojalada', 'Poll Dorset',\n",
       "       'Ronderib Afrikaner', 'Romney', 'Salz', 'Scottish Blackface',\n",
       "       'Sakiz', 'Swiss Mirror', 'Sumatra', 'Swiss White Alpine', 'Texel',\n",
       "       'Tregaon Welsh mountain', 'Valais Blacknose',\n",
       "       'Welsh Hardy Speckled Face', 'Northern Tibetan', 'Eastern Tibetan',\n",
       "       'Ovis canadensis', 'Ovis dalli', 'North American Soay', 'Argali ',\n",
       "       'Altay sheep', 'Baerchuke sheep', 'Bayinbuluke sheep',\n",
       "       'Diqing sheep', 'Guide Black Fur Sheep', 'Ganzi sheep',\n",
       "       'Large-tailed Han sheep', 'Guide Black Fur sheep',\n",
       "       'Hulun Buir sheep', 'Hetian sheep', 'Hu sheep', 'Kazakh sheep',\n",
       "       'Lop sheep', 'Hanzhong sheep', 'Minxian Black Fur sheep',\n",
       "       'Lanzhou Large-tailed sheep', 'European  mouflon ', 'Garole sheep',\n",
       "       'Sunite sheep', 'Shiping Gray sheep', 'Tengchong sheep',\n",
       "       'Tashkurgan sheep', 'Sishui Fur sheep', 'Wadi sheep',\n",
       "       'Weining sheep', 'Taihang Fur sheep', 'Tong sheep',\n",
       "       'Wuranke sheep', 'Tan sheep', 'Ujimqin sheep',\n",
       "       'Small-tailed Han sheep'], dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_WGS[\"Breed\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d58760c-4966-4613-af05-31e1ae5dd341",
   "metadata": {},
   "source": [
    "There are some breeds from WGS data that are unknown. Should I consider such animals?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f143d5b2-5ba5-4858-8088-276947e9fd3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Bioproject ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Breed</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sample location</th>\n",
       "      <th>Material</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Assay type</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>data resource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Camb0619497</td>\n",
       "      <td>ERS1171396</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.68</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Camb10114673</td>\n",
       "      <td>ERS1171397</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.92</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Camb12117423</td>\n",
       "      <td>ERS1171398</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.61</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Camb976145</td>\n",
       "      <td>ERS1171399</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.77</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Camb997502</td>\n",
       "      <td>ERS1171400</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.47</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Camb997522</td>\n",
       "      <td>ERS1171401</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.87</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Camb997550</td>\n",
       "      <td>ERS1171402</td>\n",
       "      <td>ERP015709</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.06</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IROA-B2-5037</td>\n",
       "      <td>ERS154864</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,ilbolaghi,ilbolaghi</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.01</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IROA-B2-5296</td>\n",
       "      <td>ERS239046</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,Maku,Maku</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>11.26</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>IROA-B3-5134</td>\n",
       "      <td>ERS239047</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>male</td>\n",
       "      <td>Southern Asia,Iran,khoiy,khoiy</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.31</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IROA-B4-5190</td>\n",
       "      <td>ERS154865</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,salmas,salmas</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.4</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>IROA-B5-5295</td>\n",
       "      <td>ERS154863</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>male</td>\n",
       "      <td>Southern Asia,Iran,urumie,urumie</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.32</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IROA-B6-5139</td>\n",
       "      <td>ERS239048</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,naghade,naghade</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.15</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>IROA-C3-5212</td>\n",
       "      <td>ERS154869</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>male</td>\n",
       "      <td>Southern Asia,Iran,Ghare ziadin,Ghare ziadin</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.98</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>IROA-C6-5187</td>\n",
       "      <td>ERS239049</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,oshnavie,oshnavie</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.95</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>IROA-C7-5042</td>\n",
       "      <td>ERS239050</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,naghade,naghade</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.85</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>IROA-D5-5081</td>\n",
       "      <td>ERS239051</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,azarshahr,azarshahr</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.44</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>IROA-D6-5152</td>\n",
       "      <td>ERS154866</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,ajabshir,ajabshir</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.78</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IROA-D7-5033</td>\n",
       "      <td>ERS239052</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,naghade,naghade</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.11</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>IROA-E5-5157</td>\n",
       "      <td>ERS239053</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,asko,asko</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.95</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>IROA-E6-5351</td>\n",
       "      <td>ERS239054</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,malakan,malkan</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.39</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>IROA-E7-5036</td>\n",
       "      <td>ERS239055</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>male</td>\n",
       "      <td>Southern Asia,Iran,Miandoab,miandoab</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.14</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>IROA-F10-5068</td>\n",
       "      <td>ERS239056</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>male</td>\n",
       "      <td>Southern Asia,Iran,sanandaj,sanandaj</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.6</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>IROA-F3-5142</td>\n",
       "      <td>ERS154867</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,ahar,ahar</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.74</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>IROA-F5-5051</td>\n",
       "      <td>ERS154868</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,Bostan abad,Bostan abad</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12.87</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>IROA-G3-5095</td>\n",
       "      <td>ERS239057</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,ahar,ahar</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.15</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>IROA-G4-5205</td>\n",
       "      <td>ERS154862</td>\n",
       "      <td>ERP001582</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>female</td>\n",
       "      <td>Southern Asia,Iran,niaz,Meshkin shahr</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.69</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>QTLMAMMIT1</td>\n",
       "      <td>ERS793326</td>\n",
       "      <td>ERP011062</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.38</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>QTLMAMMIT2</td>\n",
       "      <td>ERS793327</td>\n",
       "      <td>ERP011062</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>14.15</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>QTLMAMMIT3</td>\n",
       "      <td>ERS793328</td>\n",
       "      <td>ERP011062</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13.61</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>SRP015707</td>\n",
       "      <td>SRS863604</td>\n",
       "      <td>SRP055842</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Illumina Genome Analyzer II,Illumina Genome An...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>16.46</td>\n",
       "      <td>GVM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample ID Biosample ID Bioproject ID     Species Breed     Sex  \\\n",
       "10     Camb0619497   ERS1171396     ERP015709  Ovis aries     -       -   \n",
       "11    Camb10114673   ERS1171397     ERP015709  Ovis aries     -       -   \n",
       "12    Camb12117423   ERS1171398     ERP015709  Ovis aries     -       -   \n",
       "13      Camb976145   ERS1171399     ERP015709  Ovis aries     -       -   \n",
       "14      Camb997502   ERS1171400     ERP015709  Ovis aries     -       -   \n",
       "15      Camb997522   ERS1171401     ERP015709  Ovis aries     -       -   \n",
       "16      Camb997550   ERS1171402     ERP015709  Ovis aries     -       -   \n",
       "19    IROA-B2-5037    ERS154864     ERP001582  Ovis aries     -  female   \n",
       "20    IROA-B2-5296    ERS239046     ERP001582  Ovis aries     -  female   \n",
       "21    IROA-B3-5134    ERS239047     ERP001582  Ovis aries     -    male   \n",
       "22    IROA-B4-5190    ERS154865     ERP001582  Ovis aries     -  female   \n",
       "23    IROA-B5-5295    ERS154863     ERP001582  Ovis aries     -    male   \n",
       "24    IROA-B6-5139    ERS239048     ERP001582  Ovis aries     -  female   \n",
       "25    IROA-C3-5212    ERS154869     ERP001582  Ovis aries     -    male   \n",
       "26    IROA-C6-5187    ERS239049     ERP001582  Ovis aries     -  female   \n",
       "27    IROA-C7-5042    ERS239050     ERP001582  Ovis aries     -  female   \n",
       "28    IROA-D5-5081    ERS239051     ERP001582  Ovis aries     -  female   \n",
       "29    IROA-D6-5152    ERS154866     ERP001582  Ovis aries     -  female   \n",
       "30    IROA-D7-5033    ERS239052     ERP001582  Ovis aries     -  female   \n",
       "31    IROA-E5-5157    ERS239053     ERP001582  Ovis aries     -  female   \n",
       "32    IROA-E6-5351    ERS239054     ERP001582  Ovis aries     -  female   \n",
       "33    IROA-E7-5036    ERS239055     ERP001582  Ovis aries     -    male   \n",
       "34   IROA-F10-5068    ERS239056     ERP001582  Ovis aries     -    male   \n",
       "35    IROA-F3-5142    ERS154867     ERP001582  Ovis aries     -  female   \n",
       "36    IROA-F5-5051    ERS154868     ERP001582  Ovis aries     -  female   \n",
       "37    IROA-G3-5095    ERS239057     ERP001582  Ovis aries     -  female   \n",
       "38    IROA-G4-5205    ERS154862     ERP001582  Ovis aries     -  female   \n",
       "115     QTLMAMMIT1    ERS793326     ERP011062  Ovis aries     -       -   \n",
       "116     QTLMAMMIT2    ERS793327     ERP011062  Ovis aries     -       -   \n",
       "117     QTLMAMMIT3    ERS793328     ERP011062  Ovis aries     -       -   \n",
       "119      SRP015707    SRS863604     SRP055842  Ovis aries     -       -   \n",
       "\n",
       "                                  Sample location Material  \\\n",
       "10                                              -        -   \n",
       "11                                              -        -   \n",
       "12                                              -        -   \n",
       "13                                              -        -   \n",
       "14                                              -        -   \n",
       "15                                              -        -   \n",
       "16                                              -        -   \n",
       "19         Southern Asia,Iran,ilbolaghi,ilbolaghi        -   \n",
       "20                   Southern Asia,Iran,Maku,Maku        -   \n",
       "21                 Southern Asia,Iran,khoiy,khoiy        -   \n",
       "22               Southern Asia,Iran,salmas,salmas        -   \n",
       "23               Southern Asia,Iran,urumie,urumie        -   \n",
       "24             Southern Asia,Iran,naghade,naghade        -   \n",
       "25   Southern Asia,Iran,Ghare ziadin,Ghare ziadin        -   \n",
       "26           Southern Asia,Iran,oshnavie,oshnavie        -   \n",
       "27             Southern Asia,Iran,naghade,naghade        -   \n",
       "28         Southern Asia,Iran,azarshahr,azarshahr        -   \n",
       "29           Southern Asia,Iran,ajabshir,ajabshir        -   \n",
       "30             Southern Asia,Iran,naghade,naghade        -   \n",
       "31                   Southern Asia,Iran,asko,asko        -   \n",
       "32              Southern Asia,Iran,malakan,malkan        -   \n",
       "33           Southern Asia,Iran,Miandoab,miandoab        -   \n",
       "34           Southern Asia,Iran,sanandaj,sanandaj        -   \n",
       "35                   Southern Asia,Iran,ahar,ahar        -   \n",
       "36     Southern Asia,Iran,Bostan abad,Bostan abad        -   \n",
       "37                   Southern Asia,Iran,ahar,ahar        -   \n",
       "38          Southern Asia,Iran,niaz,Meshkin shahr        -   \n",
       "115                                             -    Blood   \n",
       "116                                             -    Blood   \n",
       "117                                             -    Blood   \n",
       "119                                             -        -   \n",
       "\n",
       "                                            Technology Assay type Coverage  \\\n",
       "10                                 Illumina HiSeq 2000        WGS    13.68   \n",
       "11                                 Illumina HiSeq 2000        WGS    13.92   \n",
       "12                                 Illumina HiSeq 2000        WGS    12.61   \n",
       "13                                 Illumina HiSeq 2000        WGS    12.77   \n",
       "14                                 Illumina HiSeq 2000        WGS    13.47   \n",
       "15                                 Illumina HiSeq 2000        WGS    12.87   \n",
       "16                                 Illumina HiSeq 2000        WGS    12.06   \n",
       "19                                 Illumina HiSeq 2000        WGS    14.01   \n",
       "20                                 Illumina HiSeq 2000        WGS    11.26   \n",
       "21                                 Illumina HiSeq 2000        WGS    12.31   \n",
       "22                                 Illumina HiSeq 2000        WGS     13.4   \n",
       "23                                 Illumina HiSeq 2000        WGS    13.32   \n",
       "24                                 Illumina HiSeq 2000        WGS    12.15   \n",
       "25                                 Illumina HiSeq 2000        WGS    13.98   \n",
       "26                                 Illumina HiSeq 2000        WGS    14.95   \n",
       "27                                 Illumina HiSeq 2000        WGS    13.85   \n",
       "28                                 Illumina HiSeq 2000        WGS    14.44   \n",
       "29                                 Illumina HiSeq 2000        WGS    13.78   \n",
       "30                                 Illumina HiSeq 2000        WGS    12.11   \n",
       "31                                 Illumina HiSeq 2000        WGS    12.95   \n",
       "32                                 Illumina HiSeq 2000        WGS    12.39   \n",
       "33                                 Illumina HiSeq 2000        WGS    13.14   \n",
       "34                                 Illumina HiSeq 2000        WGS     13.6   \n",
       "35                                 Illumina HiSeq 2000        WGS    13.74   \n",
       "36                                 Illumina HiSeq 2000        WGS    12.87   \n",
       "37                                 Illumina HiSeq 2000        WGS    13.15   \n",
       "38                                 Illumina HiSeq 2000        WGS    13.69   \n",
       "115                                Illumina HiSeq 2000        WGS    14.38   \n",
       "116                                Illumina HiSeq 2000        WGS    14.15   \n",
       "117                                Illumina HiSeq 2000        WGS    13.61   \n",
       "119  Illumina Genome Analyzer II,Illumina Genome An...        WGS    16.46   \n",
       "\n",
       "    data resource  \n",
       "10            GVM  \n",
       "11            GVM  \n",
       "12            GVM  \n",
       "13            GVM  \n",
       "14            GVM  \n",
       "15            GVM  \n",
       "16            GVM  \n",
       "19            GVM  \n",
       "20            GVM  \n",
       "21            GVM  \n",
       "22            GVM  \n",
       "23            GVM  \n",
       "24            GVM  \n",
       "25            GVM  \n",
       "26            GVM  \n",
       "27            GVM  \n",
       "28            GVM  \n",
       "29            GVM  \n",
       "30            GVM  \n",
       "31            GVM  \n",
       "32            GVM  \n",
       "33            GVM  \n",
       "34            GVM  \n",
       "35            GVM  \n",
       "36            GVM  \n",
       "37            GVM  \n",
       "38            GVM  \n",
       "115           GVM  \n",
       "116           GVM  \n",
       "117           GVM  \n",
       "119           GVM  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isheep_WGS[isheep_WGS[\"Breed\"] == \" -\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6cf4b46-1506-4249-91cc-9a96f44ed760",
   "metadata": {},
   "source": [
    "Some breeds are missing also in the original dataset, for example the \"*IROA\\**\" samples (which seems to come from nextgen dataset):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7ef9b1d3-c048-4905-8377-bc3479a1b0e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#sample_name</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>biosamples_id</th>\n",
       "      <th>sample_provider</th>\n",
       "      <th>species</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>breed</th>\n",
       "      <th>country</th>\n",
       "      <th>closest_city</th>\n",
       "      <th>closest_locality</th>\n",
       "      <th>estimated_age_months</th>\n",
       "      <th>sex</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>sampling_date</th>\n",
       "      <th>photographs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IROA-B2-5037</td>\n",
       "      <td>ERS154864</td>\n",
       "      <td>SAMEA2012928</td>\n",
       "      <td>NEXTGEN</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>9940</td>\n",
       "      <td>.</td>\n",
       "      <td>Iran</td>\n",
       "      <td>ilbolaghi</td>\n",
       "      <td>ilbolaghi</td>\n",
       "      <td>60</td>\n",
       "      <td>female</td>\n",
       "      <td>+44.928195</td>\n",
       "      <td>+39.044641</td>\n",
       "      <td>2011-10-24</td>\n",
       "      <td>IROA-B2-5037a.JPG,IROA-B2-5037c.JPG,IROA-B2-50...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IROA-B2-5296</td>\n",
       "      <td>ERS239046</td>\n",
       "      <td>SAMEA2065588</td>\n",
       "      <td>NEXTGEN</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>9940</td>\n",
       "      <td>.</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Maku</td>\n",
       "      <td>Maku</td>\n",
       "      <td>36</td>\n",
       "      <td>female</td>\n",
       "      <td>+44.929098</td>\n",
       "      <td>+39.045764</td>\n",
       "      <td>2011-10-27</td>\n",
       "      <td>IROA-B2-5296c.JPG,IROA-B2-5296b.JPG,IROA-B2-52...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IROA-B3-5134</td>\n",
       "      <td>ERS239047</td>\n",
       "      <td>SAMEA2065589</td>\n",
       "      <td>NEXTGEN</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>9940</td>\n",
       "      <td>.</td>\n",
       "      <td>Iran</td>\n",
       "      <td>khoiy</td>\n",
       "      <td>khoiy</td>\n",
       "      <td>24</td>\n",
       "      <td>male</td>\n",
       "      <td>+44.93919</td>\n",
       "      <td>+38.656644</td>\n",
       "      <td>2011-10-30</td>\n",
       "      <td>IROA-B3-5134c.JPG,IROA-B3-5134a.JPG,IROA-B3-51...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IROA-B4-5190</td>\n",
       "      <td>ERS154865</td>\n",
       "      <td>SAMEA2012929</td>\n",
       "      <td>NEXTGEN</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>9940</td>\n",
       "      <td>.</td>\n",
       "      <td>Iran</td>\n",
       "      <td>salmas</td>\n",
       "      <td>salmas</td>\n",
       "      <td>36</td>\n",
       "      <td>female</td>\n",
       "      <td>+44.838767</td>\n",
       "      <td>+38.153121</td>\n",
       "      <td>2011-10-27</td>\n",
       "      <td>IROA-B4-5190c.JPG,IROA-B4-5190a.JPG,IROA-B4-51...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IROA-B5-5295</td>\n",
       "      <td>ERS154863</td>\n",
       "      <td>SAMEA2012927</td>\n",
       "      <td>NEXTGEN</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>9940</td>\n",
       "      <td>.</td>\n",
       "      <td>Iran</td>\n",
       "      <td>urumie</td>\n",
       "      <td>urumie</td>\n",
       "      <td>24</td>\n",
       "      <td>male</td>\n",
       "      <td>+44.952849</td>\n",
       "      <td>+37.972674</td>\n",
       "      <td>2011-10-26</td>\n",
       "      <td>IROA-B5-5295a.JPG,IROA-B5-5295c.JPG,IROA-B5-52...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #sample_name sample_accession biosamples_id sample_provider     species  \\\n",
       "0  IROA-B2-5037        ERS154864  SAMEA2012928         NEXTGEN  Ovis aries   \n",
       "1  IROA-B2-5296        ERS239046  SAMEA2065588         NEXTGEN  Ovis aries   \n",
       "2  IROA-B3-5134        ERS239047  SAMEA2065589         NEXTGEN  Ovis aries   \n",
       "3  IROA-B4-5190        ERS154865  SAMEA2012929         NEXTGEN  Ovis aries   \n",
       "4  IROA-B5-5295        ERS154863  SAMEA2012927         NEXTGEN  Ovis aries   \n",
       "\n",
       "   taxonomy_id breed country closest_city closest_locality  \\\n",
       "0         9940     .    Iran    ilbolaghi        ilbolaghi   \n",
       "1         9940     .    Iran         Maku             Maku   \n",
       "2         9940     .    Iran        khoiy            khoiy   \n",
       "3         9940     .    Iran       salmas           salmas   \n",
       "4         9940     .    Iran       urumie           urumie   \n",
       "\n",
       "  estimated_age_months     sex   longitude    latitude sampling_date  \\\n",
       "0                   60  female  +44.928195  +39.044641    2011-10-24   \n",
       "1                   36  female  +44.929098  +39.045764    2011-10-27   \n",
       "2                   24    male   +44.93919  +38.656644    2011-10-30   \n",
       "3                   36  female  +44.838767  +38.153121    2011-10-27   \n",
       "4                   24    male  +44.952849  +37.972674    2011-10-26   \n",
       "\n",
       "                                         photographs  \n",
       "0  IROA-B2-5037a.JPG,IROA-B2-5037c.JPG,IROA-B2-50...  \n",
       "1  IROA-B2-5296c.JPG,IROA-B2-5296b.JPG,IROA-B2-52...  \n",
       "2  IROA-B3-5134c.JPG,IROA-B3-5134a.JPG,IROA-B3-51...  \n",
       "3  IROA-B4-5190c.JPG,IROA-B4-5190a.JPG,IROA-B4-51...  \n",
       "4  IROA-B5-5295a.JPG,IROA-B5-5295c.JPG,IROA-B5-52...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nextgenfile = str(get_project_dir() / \"data/external/SHE/ISHEEP/ovis.sample_index\")\n",
    "nextgen = pd.read_table(nextgenfile, skiprows=10)\n",
    "nextgen.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c653a8f-8680-4666-a810-1bed0ca56224",
   "metadata": {},
   "source": [
    "However, I could collect coordinates from this datafile"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c4c5b0f-26d6-46c5-b726-d8be48f6151b",
   "metadata": {},
   "source": [
    "Try to collect data from CNCB biosample or load data from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "68fc9237-c29f-44ee-825f-0934549b83f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if Path(\"isheep_WGS_biosample.pkl\").exists():\n",
    "    isheep_WGS_biosample = pd.read_pickle(\"isheep_WGS_biosample.pkl\")\n",
    "else:\n",
    "    biosample_data = [fetch_cncb_biosample(biosample_id) for biosample_id in tqdm(isheep_WGS[\"Biosample ID\"])]\n",
    "    isheep_WGS_biosample = pd.json_normalize(biosample_data)\n",
    "    isheep_WGS_biosample.to_pickle(\"isheep_WGS_biosample.pkl\")\n",
    "    isheep_WGS_biosample.to_excel(\"isheep_WGS_biosample.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3dd2a16d-ee1d-4be0-9021-eae132be4e17",
   "metadata": {},
   "outputs": [],
   "source": [
    "plinkio = plinkfile.open(str(get_project_dir() / \"data/external/SHE/ISHEEP/WGS-all/WGS-all.smarter\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43ab1f4c-dea2-4d54-aa69-bc6495c340d6",
   "metadata": {},
   "source": [
    "There are some samples which I can't find in cncb biosamples: They are annotated using SRA accession and so I can't fetch a record for more than 100 samples. Other samples seems to have geographic coordinates in the same format of 600K. Check that sample names in plink file and metadata are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e5d104bc-3fbe-4f3e-8717-30c977734d37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "127 samples name don't match between metadata and plink file\n"
     ]
    }
   ],
   "source": [
    "samples = [sample.iid for sample in plinkio.get_samples()]\n",
    "count = 0\n",
    "\n",
    "for index, row in isheep_WGS.iterrows():\n",
    "    if not row[\"Sample ID\"] in samples:\n",
    "        # print(f\"Missing {row['Sample ID']}\")\n",
    "        count += 1\n",
    "        \n",
    "if count > 0:\n",
    "    print(f\"{count} samples name don't match between metadata and plink file\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "06d217ee-7db8-419a-9b88-62f8ea55c54e",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.DataFrame({\"samples_id\": isheep_WGS[\"Sample ID\"], \"samples_vcf\": samples})\n",
    "if not Path(\"samples_id-vcf-WGS.xlsx\").exists():\n",
    "    test.to_excel(\"samples_id-vcf-WGS.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37744bae-47b8-4f35-90f1-6c0f45149935",
   "metadata": {},
   "source": [
    "IDs are different in metadata and datafiles. Save only WGS sample in a different file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ca72d33a-afdb-4dc5-9b4b-3670a9c2e6bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "isheep_WGS.to_excel(\"isheep_WGS.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
